lib/client.js

```diff
@@ -170,23 +211,48 @@ Client.prototype._invoke = function(method, args, location, callback, options, e
   self.lastMessage = message;
   self.lastRequest = xml;
 
-  req = http.request(location, xml, function(err, response, body) {
+  self.emit('message', message);
+  self.emit('request', xml);
+
+  req = self.httpClient.request(location, xml, function(err, response, body) {
     var result;
     var obj;
     self.lastResponse = body;
     self.lastResponseHeaders = response && response.headers;
+    self.emit('response', body);
 
     if (err) {
       callback(err);
-    } else if (response.statusCode !== 200) {
-      callback(new Error('Invalid response: ' + response.statusCode + '\nBody: ' + body));
     } else {
       try {
         obj = self.wsdl.xmlToObject(body);
+
+        if (!obj.Body && response.statusCode !== 200) {
+          return callback(new Error('Invalid response: ' + response.statusCode + '\nBody: ' + body));
+        }
       } catch (error) {
+        //When the output element cannot be looked up in the wsdl,
+        //Then instead of sending the error, We pass the body in the response.
+        if(output && !output.$lookupTypes) {
+          debug('Response element is not present. Unable to convert response xml to json.');
+          return callback(null,response,body,null);
+        }
+        error.response = response;
+        error.body = body;
+        self.emit('soapError', error);
+        return callback(error, response, body, null);
+      }
+
+      if (!output){
+        // one-way, no output expected
+        return callback(null, null, body, obj.Header);
+      }
+
+      if( typeof obj.Body !== 'object' ) {
+        var error = new Error('Cannot parse response');
         error.response = response;
         error.body = body;
-        return callback(error, response, body);
+        return callback(error, obj, body, null);
```

Comment:
I understand that we want to know what `obj` is in case of parsing error. But it kind of break the pattern where second param is `response` when there is an error, and `result` when there is none. I don't really have a better solution. Maybe add `obj` to `error`? I don't understand why they want to add `response` and `body` to `error` when they actually pass them in the callback anyway.